```
Running distributed tests for the Gloo backend
Process process 1:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "./test_distributed.py", line 836, in _run
    getattr(self, self.id().split(".")[2])()
  File "./test_distributed.py", line 792, in wrapper
    fn(self)
  File "./test_distributed.py", line 41, in wrapper
    return func(*args, **kwargs)
  File "./test_distributed.py", line 56, in wrapper
    return func(*args, **kwargs)
  File "./test_distributed.py", line 747, in test_DistributedDataParallel
    model_DDP.cuda(gpu_subset[0])
IndexError: list index out of range
Process process 2:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "./test_distributed.py", line 836, in _run
    getattr(self, self.id().split(".")[2])()
  File "./test_distributed.py", line 792, in wrapper
    fn(self)
  File "./test_distributed.py", line 41, in wrapper
    return func(*args, **kwargs)
  File "./test_distributed.py", line 56, in wrapper
    return func(*args, **kwargs)
  File "./test_distributed.py", line 747, in test_DistributedDataParallel
    model_DDP.cuda(gpu_subset[0])
IndexError: list index out of range
Process process 0:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "./test_distributed.py", line 836, in _run
    getattr(self, self.id().split(".")[2])()
  File "./test_distributed.py", line 792, in wrapper
    fn(self)
  File "./test_distributed.py", line 41, in wrapper
    return func(*args, **kwargs)
  File "./test_distributed.py", line 56, in wrapper
    return func(*args, **kwargs)
  File "./test_distributed.py", line 747, in test_DistributedDataParallel
    model_DDP.cuda(gpu_subset[0])
IndexError: list index out of range
F.s.s......s........sss.ssssssssssssssss
```

--- a/test/test_distributed.py
+++ b/test/test_distributed.py
@@ -2,6 +2,7 @@ import fcntl
 import multiprocessing
 import os
 import sys
+import copy
 import time
 import unittest
 from functools import wraps, reduce
@@ -11,7 +12,7 @@ import torch
 import torch.cuda
 import torch.nn as nn
 import torch.distributed as dist
-import torch.autograd.Variable as Variable
+from torch.autograd import Variable
 from common import TestCase
 
 BACKEND = os.environ['BACKEND']
@@ -54,6 +55,16 @@ def skip_if_no_multigpu(func):
 
         return func(*args, **kwargs)
     return wrapper
+
+
+def apply_hack_for_nccl():
+    # This is a hack for a known NCCL issue using multiprocess
+    # in conjunction with multiple threads to manage different GPUs which
+    # may cause ncclCommInitRank to fail.
+    # http://docs.nvidia.com/deeplearning/sdk/nccl-release-notes/rel_2.1.4.html#rel_2.1.4
+    # It slows down the performance of collective operations.
+    # Without this setting NCCL might throw unhandled error.
+    os.environ['NCCL_MAX_NRINGS'] = '1'
 
 
 @contextmanager
@@ -566,7 +577,7 @@ class _DistTestBase(object):
         group, group_id, rank = self._init_group_test()
         self._test_barrier_helper(group, group_id, rank)
 
-    # MULTIGPU TESTS(ONLY FOR NCCL BACKEND)
+    # MULTIGPU TESTS
     def _init_multigpu_helper(self):
         """Multigpu tests are designed to simulate the multi nodes with multi
         GPUs on each node. Nccl backend requires equal #GPUs in each process.
@@ -577,13 +588,8 @@ class _DistTestBase(object):
         world_size = dist.get_world_size()
         visible_devices = range(nGPUs)
 
-        # This is a hack for a known NCCL issue using multiprocess
-        # in conjunction with multiple threads to manage different GPUs which
-        # may cause ncclCommInitRank to fail.
-        # http://docs.nvidia.com/deeplearning/sdk/nccl-release-notes/rel_2.1.4.html#rel_2.1.4
-        # It slows down the performance of collective operations.
-        # Without this setting NCCL might throw unhandled error.
-        os.environ['NCCL_MAX_NRINGS'] = '1'
+        if BACKEND == 'nccl':
+            apply_hack_for_nccl()
 
         nGPUs_per_process = int(nGPUs / world_size)
         rankToGPUMapping = {}
@@ -713,9 +719,13 @@ class _DistTestBase(object):
 
     @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
                      "Only Nccl & Gloo backend support DistributedDataParallel")
+    @skip_if_no_cuda_distributed
+    @skip_if_no_multigpu
     def test_DistributedDataParallel(self):
+        # Run a simple end to end DDP model, use result of single node model
+        # as baseline
         group, group_id, rank = self._init_global_test()
-        os.environ['NCCL_MAX_NRINGS'] = '1'
+        rankToGPUMapping = self._init_multigpu_helper()
 
         class Net(nn.Module):
             def __init__(self):
@@ -726,21 +736,43 @@ class _DistTestBase(object):
                 x = self.features(x)
                 return x
 
+        # cpu training setup
         model = Net()
-        model.cuda()
+        model_cpu = copy.deepcopy(model)
+        optimizer_cpu = torch.optim.SGD(model_cpu.parameters(), 0.1)
 
-        nGPUs = torch.cuda.device_count()
-        input_var = Variable(torch.randn(nGPUs, 1, 2), requires_grad=True).cuda()
-        target = Variable(torch.randn(nGPUs, 1, 4)).cuda()
+        # DDP training setup
+        model_DDP = copy.deepcopy(model)
+        gpu_subset = list(rankToGPUMapping[rank])
+        model_DDP.cuda(gpu_subset[0])
+        model_DDP = nn.parallel.DistributedDataParallel(model_DDP,
+                                                        device_ids=gpu_subset)
+        optimizer_DDP = torch.optim.SGD(model_DDP.parameters(), 0.1)
+
+        # batch_size for DDP should be divisible by #GPU per node.
+        batch_size = len(gpu_subset) * int(WORLD_SIZE)
+        input_cpu = torch.randn(batch_size, 1, 2)
+        target = torch.randn(batch_size, 1, 4)
         loss = nn.MSELoss()
-        optimizer = torch.optim.SGD(model.parameters(), 0.1)
 
-        model_expected = self._test_DDP_helper(model, input_var, target, loss, optimizer)
+        model_cpu = self._test_DDP_helper(
+            model_cpu,
+            Variable(input_cpu, requires_grad=True),
+            Variable(target),
+            loss,
+            optimizer_cpu)
 
-        model = torch.nn.parallel.DistributedDataParallel(model)
-        model_DDP = self._test_DDP_helper(model, input_var, target, loss, optimizer)
+        # DDP scatters subsets of input_cpu to GPUs
+        model_DDP = self._test_DDP_helper(
+            model_DDP,
+            Variable(input_cpu, requires_grad=True),
+            Variable(target.cuda(gpu_subset[0], non_blocking=True)),
+            loss,
+            optimizer_DDP)
 
-        self.assertEqual(model_expected.features.weight, model_DDP.module.features.weight)
+        self.assertEqual(model_cpu.features.weight,
+                         model_DDP.module.features.weight)
+        self._barrier()
 
 if BACKEND == 'tcp' or BACKEND == 'gloo' or BACKEND == 'nccl':
     WORLD_SIZE = os.environ['WORLD_SIZE']
@@ -748,7 +780,7 @@ if BACKEND == 'tcp' or BACKEND == 'gloo' or BACKEND == 'nccl':
     class TestDistBackend(TestCase, _DistTestBase):
 
         MANAGER_PROCESS_RANK = -1
-        JOIN_TIMEOUT = 15
+        JOIN_TIMEOUT = 18
 
         @staticmethod
         def manager_join(fn):

--- a/test/test_distributed.py
+++ b/test/test_distributed.py
@@ -9,7 +9,9 @@ from contextlib import contextmanager
 
 import torch
 import torch.cuda
+import torch.nn as nn
 import torch.distributed as dist
+import torch.autograd.Variable as Variable
 from common import TestCase
 
 BACKEND = os.environ['BACKEND']
@@ -699,6 +701,46 @@ class _DistTestBase(object):
         rankToGPUMapping = self._init_multigpu_helper()
         self._test_all_gather_multigpu_helper(group, group_id, rank,
                                               rankToGPUMapping)
+
+    # END TO END TEST FOR DISTRIBUTEDDATAPARALLEL
+    def _test_DDP_helper(self, model, input_var, target, loss, optimizer):
+        model.train()
+        output = model(input_var)
+        l = loss(output, target)
+        l.backward()
+        optimizer.step()
+        return model
+
+    @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
+                     "Only Nccl & Gloo backend support DistributedDataParallel")
+    def test_DistributedDataParallel(self):
+        group, group_id, rank = self._init_global_test()
+        os.environ['NCCL_MAX_NRINGS'] = '1'
+
+        class Net(nn.Module):
+            def __init__(self):
+                super(Net, self).__init__()
+                self.features = nn.Linear(2, 4, bias=False)
+
+            def forward(self, x):
+                x = self.features(x)
+                return x
+
+        model = Net()
+        model.cuda()
+
+        nGPUs = torch.cuda.device_count()
+        input_var = Variable(torch.randn(nGPUs, 1, 2), requires_grad=True).cuda()
+        target = Variable(torch.randn(nGPUs, 1, 4)).cuda()
+        loss = nn.MSELoss()
+        optimizer = torch.optim.SGD(model.parameters(), 0.1)
+
+        model_expected = self._test_DDP_helper(model, input_var, target, loss, optimizer)
+
+        model = torch.nn.parallel.DistributedDataParallel(model)
+        model_DDP = self._test_DDP_helper(model, input_var, target, loss, optimizer)
+
+        self.assertEqual(model_expected.features.weight, model_DDP.module.features.weight)
 
 if BACKEND == 'tcp' or BACKEND == 'gloo' or BACKEND == 'nccl':
     WORLD_SIZE = os.environ['WORLD_SIZE']

--- a/test/test_distributed.py
+++ b/test/test_distributed.py
@@ -28,7 +28,7 @@ if not dist.is_available():
 
 SKIP_IF_NO_CUDA_EXIT_CODE = 75
 SKIP_IF_NO_MULTIGPU_EXIT_CODE = 76
-
+SKIP_IF_SMALL_WORLDSIZE_EXIT_CODE = 77
 
 def skip_if_no_cuda_distributed(func):
     func.skip_if_no_cuda_distributed = True
@@ -56,6 +56,16 @@ def skip_if_no_multigpu(func):
         return func(*args, **kwargs)
     return wrapper
 
+def skip_if_small_worldsize(func):
+    func.skip_if_small_worldsize = True
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        if int(os.environ['WORLD_SIZE']) <= 2:
+            sys.exit(SKIP_IF_SMALL_WORLDSIZE_EXIT_CODE)
+
+        return func(*args, **kwargs)
+    return wrapper
 
 def apply_hack_for_nccl():
     # This is a hack for a known NCCL issue using multiprocess
@@ -286,6 +296,7 @@ class _DistTestBase(object):
         self._test_broadcast_helper(group, group_id, rank, True)
 
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_broadcast_group(self):
         group, group_id, rank = self._init_group_test()
         self._test_broadcast_helper(group, group_id, rank)
@@ -352,6 +363,7 @@ class _DistTestBase(object):
 
     @unittest.skipIf(BACKEND == 'gloo', "Gloo does not support reduce")
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_reduce_group_sum(self):
         group, group_id, rank = self._init_group_test()
         self._test_reduce_helper(
@@ -360,6 +372,7 @@ class _DistTestBase(object):
 
     @unittest.skipIf(BACKEND == 'gloo', "Gloo does not support reduce")
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_reduce_group_product(self):
         group, group_id, rank = self._init_group_test()
         self._test_reduce_helper(
@@ -369,6 +382,7 @@ class _DistTestBase(object):
 
     @unittest.skipIf(BACKEND == 'gloo', "Gloo does not support reduce")
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_reduce_group_min(self):
         group, group_id, rank = self._init_group_test()
         self._test_reduce_helper(
@@ -377,6 +391,7 @@ class _DistTestBase(object):
 
     @unittest.skipIf(BACKEND == 'gloo', "Gloo does not support reduce")
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_reduce_group_max(self):
         group, group_id, rank = self._init_group_test()
         self._test_reduce_helper(
@@ -441,6 +456,7 @@ class _DistTestBase(object):
         )
 
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_all_reduce_group_sum(self):
         group, group_id, rank = self._init_group_test()
         self._test_all_reduce_helper(
@@ -448,6 +464,7 @@ class _DistTestBase(object):
         )
 
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_all_reduce_group_product(self):
         group, group_id, rank = self._init_group_test()
         self._test_all_reduce_helper(
@@ -456,6 +473,7 @@ class _DistTestBase(object):
         )
 
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_all_reduce_group_min(self):
         group, group_id, rank = self._init_group_test()
         self._test_all_reduce_helper(
@@ -463,6 +481,7 @@ class _DistTestBase(object):
         )
 
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_all_reduce_group_max(self):
         group, group_id, rank = self._init_group_test()
         self._test_all_reduce_helper(
@@ -488,6 +507,7 @@ class _DistTestBase(object):
 
     @unittest.skipIf(BACKEND == 'gloo', "Gloo does not support scatter")
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support scatter")
+    @skip_if_small_worldsize
     def test_scatter_group(self):
         group, group_id, rank = self._init_group_test()
         self._test_scatter_helper(group, group_id, rank)
@@ -513,6 +533,7 @@ class _DistTestBase(object):
 
     @unittest.skipIf(BACKEND == 'gloo', "Gloo does not support gather")
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_gather_group(self):
         group, group_id, rank = self._init_group_test()
         self._test_gather_helper(group, group_id, rank)
@@ -545,6 +566,7 @@ class _DistTestBase(object):
         self._test_all_gather_helper(group, group_id, rank, True)
 
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_all_gather_group(self):
         group, group_id, rank = self._init_group_test()
         self._test_all_gather_helper(group, group_id, rank)
@@ -573,6 +595,7 @@ class _DistTestBase(object):
         self._test_barrier_helper(group, group_id, rank)
 
     @unittest.skipIf(BACKEND == 'nccl', "Nccl does not support newGroup")
+    @skip_if_small_worldsize
     def test_barrier_group(self):
         group, group_id, rank = self._init_group_test()
         self._test_barrier_helper(group, group_id, rank)
@@ -838,7 +861,8 @@ if BACKEND == 'tcp' or BACKEND == 'gloo' or BACKEND == 'nccl':
 
         def _join_and_reduce(self, fn):
             skip_ok = getattr(fn, "skip_if_no_cuda_distributed", False) \
-                or getattr(fn, "skip_if_no_multigpu", False)
+                or getattr(fn, "skip_if_no_multigpu", False) \
+                or getattr(fn, "skip_if_small_worldsize", False)
             for p in self.processes:
                 p.join(self.JOIN_TIMEOUT)
                 if not skip_ok:
@@ -850,7 +874,8 @@ if BACKEND == 'tcp' or BACKEND == 'gloo' or BACKEND == 'nccl':
                 # mismatched exit codes if the first isn't valid
                 assert first_process.exitcode == 0 \
                     or first_process.exitcode == SKIP_IF_NO_CUDA_EXIT_CODE \
-                    or first_process.exitcode == SKIP_IF_NO_MULTIGPU_EXIT_CODE
+                    or first_process.exitcode == SKIP_IF_NO_MULTIGPU_EXIT_CODE \
+                    or first_process.exitcode == SKIP_IF_SMALL_WORLDSIZE_EXIT_CODE
 
                 for p in self.processes:
                     self.assertEqual(p.exitcode, first_process.exitcode)
@@ -858,6 +883,8 @@ if BACKEND == 'tcp' or BACKEND == 'gloo' or BACKEND == 'nccl':
                     raise unittest.SkipTest("cuda is not available")
                 if first_process.exitcode == SKIP_IF_NO_MULTIGPU_EXIT_CODE:
                     raise unittest.SkipTest("multigpu is not available")
+                if first_process.exitcode == SKIP_IF_SMALL_WORLDSIZE_EXIT_CODE:
+                    raise unittest.SkipTest("worldsize is too small to run group tests")
 
 elif BACKEND == 'mpi':
     dist.init_process_group(init_method=INIT_METHOD, backend='mpi')

